File: Chapter04/Activity4_02/Activity4_02_DF.scala
package Chapter04.Activity4_01

import org.apache.spark.rdd.RDD
import org.apache.spark.sql.{DataFrame, SparkSession}
import Utilities01.HelperScala.createSession
import Utilities02.HelperScala.{extractRawRecords, parseRawWarc, parseRawWet}
import Utilities02.{WarcRecord, WetRecord}

object Activity4_02_DF {
  def main(args: Array[String]): Unit = {
    val inputLocWarc = "/Users/a/CC-MAIN-20191013195541-20191013222541-00000.warc" // ToDo: Modify Path
    val inputLocWet = "/Users/a/CC-MAIN-20191013195541-20191013222541-00000.warc.wet" // ToDo: Modify Path
    implicit val session: SparkSession = createSession(3, "Activity 4.2 DataFrame")

    val warcRecords: RDD[WarcRecord] = extractRawRecords(inputLocWarc).flatMap(parseRawWarc)
    val wetRecords: RDD[WetRecord] = extractRawRecords(inputLocWet).flatMap(parseRawWet)

    import session.implicits._
    val warcRecordsDf: DataFrame = warcRecords.toDF.select("targetURI", "language")
    val wetRecordsDf: DataFrame = wetRecords.toDF().select("targetURI", "plainText")

    val joinedDf = warcRecordsDf.join(wetRecordsDf, Seq("targetURI"))
    val spanishRecords = joinedDf.filter('language === "es") // median shuffle read 32.7 MB / 22412

    println(spanishRecords.count())
    Thread.sleep(1000L * 60L * 100L) // For exploring WebUI
  }
}

File: Chapter04/Activity4_02/Activity4_02_DF.py
import time
from pyspark import RDD
from pyspark.sql import SparkSession, DataFrame
from Chapter01.utilities01_py.helper_python import create_session
from Chapter02.utilities02_py.helper_python import extract_raw_records, parse_raw_warc, parse_raw_wet


if __name__ == "__main__":
    input_loc_warc = '/Users/a/CC-MAIN-20191013195541-20191013222541-00000.warc'  # ToDo: Modify path
    input_loc_wet = '/Users/a/CC-MAIN-20191013195541-20191013222541-00000.warc.wet'  # ToDo: Modify path
    session: SparkSession = create_session(3, 'Activity 4.2 DataFrame')

    warc_records: RDD = extract_raw_records(input_loc_warc, session).flatMap(lambda record: parse_raw_warc(record))
    wet_records: RDD = extract_raw_records(input_loc_wet, session).flatMap(lambda record: parse_raw_wet(record))

    from pyspark.sql.functions import col
    warc_records_df: DataFrame = warc_records.toDF().select(col('target_uri'), col('language'))
    wet_records_df: DataFrame = wet_records.toDF().select(col('target_uri'), col('plain_text'))

    joined_df = warc_records_df.join(wet_records_df, ['target_uri'])
    spanish_records = joined_df.filter(col('language') == 'es')

    time.sleep(10 * 60)  # For exploring WebUI
